### ISSUE TITLE
Compile avoidance should respect public constant changes

### ISSUE THREAD

**Issue by oehme** [gradle/gradle#1476]
The API extractor used by the new `@CompileClasspath` snapshotter drops public constants, making incremental builds incorrect. This affects the following use case:

```
class A {
  public static final int FOO = 1;
}

class B {
  public void bar() {
    System.out.println(A.FOO);
  }
}
```

When A.Foo is changed, B should be recompiled, but isn't, resulting in wrong results.

**Comment by melix** [comment]
To be clear, this happens independently of incremental compilation and only if the 2 classes are in different projects.

**Comment by oehme** [comment]
The only workarounds are calling `clean` or `--rerun-tasks` or making a dummy change to `B`.

**Comment by melix** [comment]
Here's the tentative fix. It coincidentally fixes #1474 when used with JDK 8 (but not JDK 7).
